broadcaster/backlog/backlog.go

```diff
@@ -0,0 +1,399 @@
+		return
+	}
+
+	if confirmed < head.Start() {
+		return
+	}
+
+	if confirmed > tail.End() {
+		log.Error("confirmed sequence number is past the end of stored messages", "confirmed sequence number", confirmed, "last stored sequence number", tail.end.Load())
+		b.reset()
+		return
+	}
+
+	// find the segment containing the confirmed message
+	found, err := b.Lookup(confirmed)
+	if err != nil {
+		log.Error(fmt.Sprintf("%s: clearing backlog", err.Error()))
+		b.reset()
+		return
+	}
+	segment, ok := found.(*backlogSegment)
+	if !ok {
+		log.Error("error in backlogSegment type assertion: clearing backlog")
+		b.reset()
+		return
+	}
+
+	// delete messages from the segment with the confirmed message
+	newHead := segment
+	start := head.Start()
+	if segment.End() == confirmed {
+		found = segment.Next()
+		newHead, ok = found.(*backlogSegment)
+		if !ok {
+			log.Error("error in backlogSegment type assertion: clearing backlog")
+			b.reset()
+			return
+		}
+	} else {
+		err = segment.delete(confirmed)
+		if err != nil {
+			log.Error(fmt.Sprintf("%s: clearing backlog", err.Error()))
+			b.reset()
+			return
+		}
+	}
+
+	// tidy up lookup, count and head
+	b.removeFromLookup(start, confirmed)
+	count := b.Count() + start - confirmed - uint64(1)
+	b.messageCount.Store(count)
+	b.head.Store(newHead)
+}
+
+// removeFromLookup removes all entries from the head segment's start index to
+// the given confirmed index.
+func (b *backlog) removeFromLookup(start, end uint64) {
+	b.lookupLock.Lock()
+	defer b.lookupLock.Unlock()
+	for i := start; i <= end; i++ {
+		delete(b.lookupByIndex, i)
+	}
+}
+
+// Lookup attempts to find the backlogSegment storing the given message index.
+func (b *backlog) Lookup(i uint64) (BacklogSegment, error) {
+	b.lookupLock.RLock()
+	segment, ok := b.lookupByIndex[i]
+	b.lookupLock.RUnlock()
+	if !ok {
+		return nil, fmt.Errorf("error finding backlog segment containing message with SequenceNumber %d", i)
+	}
+
+	return segment, nil
+}
+
+// Count returns the number of messages stored within the backlog.
+func (s *backlog) Count() uint64 {
+	return s.messageCount.Load()
+}
+
+// reset removes all segments from the backlog.
+func (b *backlog) reset() {
+	b.lookupLock.Lock()
+	defer b.lookupLock.Unlock()
+	b.head = atomic.Pointer[backlogSegment]{}
+	b.tail = atomic.Pointer[backlogSegment]{}
+	b.lookupByIndex = map[uint64]*backlogSegment{}
+	b.messageCount.Store(0)
+}
+
+// BacklogSegment defines the interface for backlogSegment.
+type BacklogSegment interface {
+	Start() uint64
+	End() uint64
+	Next() BacklogSegment
+	Contains(uint64) bool
+	Messages() []*m.BroadcastFeedMessage
+}
+
+// backlogSegment stores messages up to a limit defined by the backlog. It also
+// points to the next backlogSegment in the list.
+type backlogSegment struct {
+	start           atomic.Uint64
+	end             atomic.Uint64
+	messagesLock    sync.RWMutex
+	messages        []*m.BroadcastFeedMessage
+	messageCount    atomic.Uint64
```

Comment:
Since we have `messagesLock`, I think we should just replace messageCount with acquiring a read messagesLock and then checking `len(messages)`. We could do the same for `start` and `end`. I don't think the performance outweighs the risk of bugs due to mismatches between these variables and the actual messages array.

My original thinking with these being atomic was that we theoretically wouldn't need the messagesLock, but we have it now, and it seems safer anyways to have it.